File: libnd4j/include/ops/declarable/generic/sparse/dense_to_csr_bp.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// dense_to_csr_bp — backward pass for dense_to_csr.
//
// Forward:  dense[rows, cols] (TArgs: threshold)
//              → values[nnz], colIdx[nnz], rowPtr[rows+1]
//
// Backward: scatter gradValues back into a zeroed dDense matrix.
//
// Inputs:
//   [0] colIdx     [nnz],     INT32 — CSR column indices (forward output[1])
//   [1] rowPtr     [rows+1],  INT32 — CSR row pointers  (forward output[2])
//   [2] gradValues [nnz],     float dtype — upstream gradient w.r.t. values output
// IArgs:
//   [0] rows
//   [1] cols
// Output:
//   [0] dDense [rows, cols], same float dtype as gradValues — gradient w.r.t. dense input
//       (OUTPUT_NULLIFIED — framework pre-zeros the output; only non-zero positions are written)
//
// Math:  dDense[i, colIdx[e]] = gradValues[e]  for each entry e.
// In a valid CSR pattern every (i, colIdx[e]) pair is unique → no atomics.
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_dense_to_csr_bp)

#include <ops/declarable/headers/sparse.h>
#include <ops/declarable/helpers/sparse_convert_bp.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(dense_to_csr_bp, 3, 1, false, 0, 2) {
  auto colIdx     = INPUT_VARIABLE(0);
  auto rowPtr     = INPUT_VARIABLE(1);
  auto gradValues = INPUT_VARIABLE(2);

  const LongType rows = INT_ARG(0);
  const LongType cols = INT_ARG(1);

  REQUIRE_TRUE(colIdx->rankOf() == 1, 0,
               "dense_to_csr_bp: colIdx must be 1D, got rank %d", colIdx->rankOf());
  REQUIRE_TRUE(rowPtr->rankOf() == 1, 0,
               "dense_to_csr_bp: rowPtr must be 1D, got rank %d", rowPtr->rankOf());
  REQUIRE_TRUE(gradValues->rankOf() == 1, 0,
               "dense_to_csr_bp: gradValues must be 1D, got rank %d", gradValues->rankOf());
  REQUIRE_TRUE(rowPtr->lengthOf() == rows + 1, 0,
               "dense_to_csr_bp: rowPtr length must be rows+1=%lld, got %lld",
               (long long)(rows + 1), (long long)rowPtr->lengthOf());
  REQUIRE_TRUE(colIdx->lengthOf() == gradValues->lengthOf(), 0,
               "dense_to_csr_bp: colIdx and gradValues must have the same length");
  REQUIRE_TRUE(colIdx->dataType() == rowPtr->dataType(), 0,
               "dense_to_csr_bp: colIdx and rowPtr must have the same integer dtype");

  auto dDense = OUTPUT_NULLIFIED(0);

  const LongType nnz = colIdx->lengthOf();
  if (nnz > 0) {
    sd::ops::helpers::dense_to_csr_bp(*colIdx, *rowPtr, *gradValues, *dDense, rows, cols);
  }
  // If nnz == 0, dDense stays all-zero (already provided by OUTPUT_NULLIFIED).

  return sd::Status::OK;
}

DECLARE_SHAPE_FN(dense_to_csr_bp) {
  auto gradValues = INPUT_VARIABLE(2);

  const LongType rows = INT_ARG(0);
  const LongType cols = INT_ARG(1);

  // dDense has shape [rows, cols] with the same float dtype as gradValues
  auto outShape = ConstantShapeHelper::getInstance().createShapeInfo(
      gradValues->dataType(), 'c', {rows, cols});
  return SHAPELIST(outShape);
}

DECLARE_TYPES(dense_to_csr_bp) {
  getOpDescriptor()
      ->setAllowedInputTypes(0, {ALL_INTS})     // colIdx
      ->setAllowedInputTypes(1, {ALL_INTS})     // rowPtr
      ->setAllowedInputTypes(2, {ALL_FLOATS})   // gradValues
      ->setAllowedOutputTypes(0, {ALL_FLOATS}); // dDense
}

}  // namespace ops
}  // namespace sd

#endif  // NOT_EXCLUDED(OP_dense_to_csr_bp)

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/sparse/DenseToCsrBp.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.sparse;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Collections;
import java.util.List;

/**
 * Backward pass for {@link DenseToCsr}: scatter upstream gradients from the CSR values output
 * back into the dense gradient matrix.
 *
 * <p>C++ op name: {@code dense_to_csr_bp}
 *
 * <p><b>Inputs (3 arrays):</b>
 * <ol>
 *   <li>{@code colIdx}     – 1D [nnz], INT32 — CSR column indices (forward output[1])</li>
 *   <li>{@code rowPtr}     – 1D [rows+1], INT32 — CSR row pointers (forward output[2])</li>
 *   <li>{@code gradValues} – 1D [nnz], float — upstream gradient w.r.t. the values output</li>
 * </ol>
 *
 * <p><b>Integer arguments (IArgs):</b> {@code rows}, {@code cols}
 *
 * <p><b>Output:</b>
 * <ol>
 *   <li>{@code dDense} – 2D [rows, cols], same float dtype as {@code gradValues} —
 *       gradient w.r.t. the dense input (pre-zeroed by the framework)</li>
 * </ol>
 *
 * <p>Math: {@code dDense[i, colIdx[e]] = gradValues[e]} for each entry {@code e}.
 * In a valid CSR pattern every (row, col) pair is unique, so no atomics are needed.
 *
 * <p>This op is forward-only (no {@code doDiff}); it is the gradient leaf for
 * {@link DenseToCsr#doDiff}.
 */
public class DenseToCsrBp extends DynamicCustomOp {

    /** No-arg constructor required for ImportClassMapping reflection. */
    public DenseToCsrBp() {}

    /**
     * Eager (INDArray) constructor.
     *
     * @param colIdx     1D [nnz] INT32 — CSR column indices (from forward output)
     * @param rowPtr     1D [rows+1] INT32 — CSR row pointers (from forward output)
     * @param gradValues 1D [nnz] float — upstream gradient w.r.t. the values output
     * @param rows       number of rows in the original dense matrix
     * @param cols       number of columns in the original dense matrix
     */
    public DenseToCsrBp(INDArray colIdx, INDArray rowPtr, INDArray gradValues,
                         long rows, long cols) {
        super(new INDArray[]{colIdx, rowPtr, gradValues}, null);
        addIArgument(rows, cols);
    }

    /**
     * SameDiff (symbolic) constructor.
     *
     * @param sd         SameDiff instance
     * @param colIdx     SDVariable [nnz] INT32 — CSR column indices
     * @param rowPtr     SDVariable [rows+1] INT32 — CSR row pointers
     * @param gradValues SDVariable [nnz] float — upstream gradient w.r.t. the values output
     * @param rows       number of rows
     * @param cols       number of columns
     */
    public DenseToCsrBp(SameDiff sd, SDVariable colIdx, SDVariable rowPtr,
                         SDVariable gradValues, long rows, long cols) {
        super(sd, new SDVariable[]{colIdx, rowPtr, gradValues});
        addIArgument(rows, cols);
    }

    @Override
    public String opName() {
        return "dense_to_csr_bp";
    }

    /**
     * Output data type: same float dtype as {@code gradValues} (input index 2).
     */
    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
        return Collections.singletonList(dataTypes.get(2));
    }
}
